rocAL/rocAL/source/pipeline/tensor.cpp

```diff
@@ -0,0 +1,357 @@
+    vx_size stride[num_of_dims];
+    void *ptr[1] = {nullptr};
+
+    stride[0] = tensor_data_size(_info.data_type());
+    for(unsigned i = 1; i < num_of_dims; i++)
+        stride[i] = stride[i - 1] * _info.dims().at(i - 1);
+
+    _vx_handle = vxCreateTensorFromHandle(_context, _info.num_of_dims(), _info.dims().data(), tensor_data_type, 0, stride, ptr, vx_mem_type(_info._mem_type));
+    vx_status status;
+    if ((status = vxGetStatus((vx_reference)_vx_handle)) != VX_SUCCESS)
+        THROW("Error: vxCreateTensorFromHandle(input: failed " + TOSTR(status))
+    _info._type = rocALTensorInfo::Type::HANDLE;
+    return 0;
+}
+
+int rocALTensor::create(vx_context context)
+{
+    if (_vx_handle)
+        return -1;
+
+    _context = context;
+
+    vx_status status;
+    vx_enum tensor_data_type = interpret_tensor_data_type(_info.data_type());
+    _vx_handle = vxCreateTensor(context, _info.num_of_dims(), _info.dims().data(), tensor_data_type, 0);
+    if ((status = vxGetStatus((vx_reference)_vx_handle)) != VX_SUCCESS)
+        THROW("Error: vxCreateTensor(input: failed " + TOSTR(status))
+    _info._type = rocALTensorInfo::Type::REGULAR;
+    return 0;
+}
+
+#if ENABLE_OPENCL
+unsigned rocALTensor::copy_data(cl_command_queue queue, unsigned char *user_buffer, bool sync)
+{
+    if (_info._type != rocALTensorInfo::Type::HANDLE)
+        return 0;
+
+    if (_info._mem_type == RocalMemType::OCL)
+    {
+        cl_int status;
+        if ((status = clEnqueueReadBuffer(queue,
+                                          (cl_mem)_mem_handle,
+                                          sync ? (CL_TRUE) : CL_FALSE,
+                                          0,
+                                          _info.data_size(),
+                                          user_buffer,
+                                          0, nullptr, nullptr)) != CL_SUCCESS)
+            THROW("clEnqueueReadBuffer failed: " + TOSTR(status))
+    }
+    else
+    {
+        memcpy(user_buffer, _mem_handle, _info.data_size());
+    }
+    return 0;
+}
+#elif ENABLE_HIP
+unsigned rocALTensor::copy_data(hipStream_t stream, void* host_memory, bool sync)
+{
+    if(_info._type != rocALTensorInfo::Type::HANDLE)
+        return 0;
+
+    if (_info._mem_type == RocalMemType::HIP)
+    {
+        // copy from device to host
+        hipError_t status;
+        if ((status = hipMemcpyDtoHAsync((void *)host_memory, _mem_handle, _info.data_size(), stream)))
+            THROW("copy_data::hipMemcpyDtoH failed: " + TOSTR(status))
+        if (sync)
+        {
+            if ((status =hipStreamSynchronize(stream)))
+                THROW("copy_data::hipStreamSynchronize failed: " + TOSTR(status))
+        }
+    }
+    else
+    {
+        memcpy(host_memory, _mem_handle, _info.data_size());
+    }
+    return 0;
+}
+#endif
+
+unsigned rocALTensor::copy_data(void* user_buffer)
+{
+    if(_info._type != rocALTensorInfo::Type::HANDLE)
+        return 0;
+
+#if ENABLE_HIP
```

Comment:
Is there an equivalent implementation for OpenCL

Reply:
In AMD OpenVX extensions we do not have the OpenCL support for tensor augmentations, as per our previous discussions  we have only HIP or HOST support hence we have not added the OpenCL copy implementation.